### PR TITLE
Follow up to #2401

### PR DESCRIPTION
**Pull request by JonRowe**
Move the logic for detecting files vs streams back to formatter loader, this restores the deprecation stream behaviour and simplifies config.

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -4,6 +4,8 @@ Enhancements:
 
 * Add seed information to JSON formatter output. (#2388, Mitsutaka Mimura)
 * Include example id in the JSON formatter output. (#2369, Xavier Shay)
+* Allow changing `output_stream` after formatters have been setup and have
+  formatter output follow. (#2401, #2419, Ilya Lavrov)
```
Comment (myronmarston) [inline Changelog.md]: "after formatters have been setup and have formatter output follow" is worded a bit awkwardly, IMO. WDYT about this instead?

> Respect changes to `config.output_stream` after formatters have been instantiated. (#2401, #2419, Ilya Lavrov)

File: lib/rspec/core/configuration.rb [hunk 2]
```diff
@@ -155,7 +155,7 @@ def deprecation_stream=(value)
             "it to take effect, or use the `--deprecation-out` CLI option. " \
             "(Called from #{CallerFilter.first_non_rspec_line})"
```
Comment (myronmarston) [inline lib/rspec/core/configuration.rb]: This warning seems like it's a different way to deal with the same issue that we're addressing for `output_stream`: what to do when users change `config.deprecation_stream` after the reporter has been initialized.  Could we apply the `OutputWrapper` solution to this config setting as well, and remove the warning?  Hopefully that's not much work...

File: lib/rspec/core/configuration.rb [hunk 3]
```diff
@@ -142,7 +142,7 @@ def expose_dsl_globally=(value)
 
       # Determines where deprecation warnings are printed.
       # Defaults to `$stderr`.
-      # @return [IO] IO to write to
+      # @return [IO, String] IO to write to or filename to write to
```
Comment (myronmarston) [inline lib/rspec/core/configuration.rb]: How about "IO or filename to write to" instead of "IO to write to or filename to write to" ?

File: lib/rspec/core/configuration.rb [hunk 4]
```diff
@@ -473,8 +473,7 @@ def initialize
       # Used to set higher priority option values from the command line.
       def force(hash)
         ordering_manager.force(hash)
-
-        @preferred_options.merge!(hash).merge!(prepared_streams(hash))
+        @preferred_options.merge!(hash)
 
         return unless hash.key?(:example_status_persistence_file_path)
         clear_values_derived_from_example_status_persistence_file_path
```
Comment (myronmarston) [inline lib/rspec/core/configuration.rb]: WDYT of changing `reset` so that it delegates to `reset_reporter` to reset the 3 ivars that method takes care of, instead of manually resetting those same 3 ivars?

File: lib/rspec/core/formatters.rb [hunk 5]
```diff
@@ -249,5 +251,15 @@ def underscore(camel_cased_word)
       word.downcase!
       word
     end
+
+    def file_at(path)
```
Comment (myronmarston) [inline lib/rspec/core/formatters.rb]: It's a little weird for the arg to be called `path` when it could be an `OutputWrapper` (which is definitely not a path!).  Also, I don't think `file_at` is the best name anymore given in the wrapper case it doesn't return a file. 
 Finally, instead of assuming it's a string in the non-wrapper case, I'd prefer we use a `case` statement for both to make fewer unproven assumptions.  Putting those together:

``` ruby
def open_file(path_or_output_wrapper)
  case path_or_output_wrapper
  when RSpec::Core::OutputWrapper
    path_or_output_wrapper = file_at(path_or_output_wrapper.output)
    path_or_output_wrapper
  when String
    RSpec::Support::DirectoryMaker.mkdir_p(File.dirname(path_or_output_wrapper))
    File.new(path_or_output_wrapper, 'w')
  end
end
```

The one thing I'm not sure on is if we should have an `else` that raises an explicit error, or just let it be `nil`.

Thoughts?

File: lib/rspec/core/output_wrapper.rb [hunk 6]
```diff
@@ -3,13 +3,17 @@ module Core
     # @private
     class OutputWrapper
       # @private
-      attr_writer :output
+      attr_accessor :output
```
Comment (myronmarston) [inline lib/rspec/core/output_wrapper.rb]: Given you changed this, should we change `@output` below to `output`?

File: lib/rspec/core/output_wrapper.rb [hunk 7]
```diff
@@ -3,13 +3,17 @@ module Core
...
 
       # @private
       def initialize(output)
         @output = output
       end
 
+      def respond_to?(name, priv=false)
+        @output.respond_to?(name, priv)
+      end
```
Comment (myronmarston) [inline lib/rspec/core/output_wrapper.rb]: Why is this needed?  We're not using `method_missing` here. All the `output` methods that we proxy are defined at load time below via `define_method`.  So actually, I think this creates a bug: if `@output` is a kind of IO object that has extra methods not defined as an `IO` instance method (such as if `@output` is a `StringIO`--in that case it has an extra `string` method), then this will make it claim it responds to the extra methods, but will not proxy them because we haven't used `define_method` for them.

File: spec/integration/output_stream_spec.rb [hunk 8]
```diff
@@ -17,7 +17,28 @@
             true
           end
         end
-      "
+      SPEC
```
Comment (myronmarston) [inline spec/integration/output_stream_spec.rb]: Why the change to the heredoc?  We haven't done that for our other integration specs.

File: spec/rspec/core/formatters_spec.rb [hunk 9]
```diff
@@ -106,6 +106,21 @@ module RSpec::Core::Formatters
...
 
+      context "with a 2nd arg defining the output" do
+        it "creates a file at that path and sets it as the output" do
+          loader.add('doc', path)
+          expect(loader.formatters.first.output).to be_a(File)
+          expect(loader.formatters.first.output.path).to eq(path)
+        end
+
+        it "accepts Pathname objects for file paths" do
+          pathname = Pathname.new(path)
```
Comment (myronmarston) [inline spec/rspec/core/formatters_spec.rb]: Do we document that `Pathname` objects are supported?  I see `IO` and `String` in our docs but hadn't noticed `Pathname`.

Also, I guess this means if you use my `case`-based implementation for the `file_at` method you'll have to make it handle `Pathname`, too.

File: spec/integration/output_stream_spec.rb [hunk 10]
```diff
@@ -17,7 +17,28 @@
...
+          c.formatter = :documentation
+          c.output_stream = 'saved_output'
+        end
+
+        RSpec.describe 'something' do
+          it 'succeeds' do
+            true
+          end
+        end
+      SPEC
```
Comment (myronmarston) [inline spec/integration/output_stream_spec.rb]: Does it make sense to add a spec for `Pathname` as well, given that's apparently supported (based on your spec below)?

File: lib/rspec/core/output_wrapper.rb (hunk 7 above)
Comment (JonRowe) [inline lib/rspec/core/output_wrapper.rb]: Our IO vs string detection depends on checking if the output responds to puts, I'm happy to define method_missing as well for the edge case, the defined methods should just be more performant.

File: spec/integration/output_stream_spec.rb (hunk 8 above)
Comment (JonRowe) [inline spec/integration/output_stream_spec.rb]: Highlighting was just giving me greif

File: Changelog.md (hunk 1 above)
Comment (JonRowe) [inline Changelog.md]: Sure, with a minor tweak of `s/instantiated/setup` as its the setup phase thats important.

File: lib/rspec/core/configuration.rb (hunk 2 above)
Comment (JonRowe) [inline lib/rspec/core/configuration.rb]: I tried, but weirdness happened, I might have time to give it another go, but maybe as another follow up pr so it doesn't block us releasing 3.6

File: lib/rspec/core/configuration.rb (hunk 4 above)
Comment (JonRowe) [inline lib/rspec/core/configuration.rb]: Sure.

File: lib/rspec/core/formatters.rb (hunk 5 above)
Comment (JonRowe) [inline lib/rspec/core/formatters.rb]: I've improved the wording, but not the implementation, the problem is `Pathname` which behaves like a string and is returned from Rails things (which is why I believe we test for it at all) but we can't assume it's loaded etc, the existing behaviour is if we can write to the possible stream, return it, otherwise assume it's a filename and open it. I'd rather preserve that behaviour.

File: spec/rspec/core/formatters_spec.rb (hunk 9 above)
Comment (JonRowe) [inline spec/rspec/core/formatters_spec.rb]: It's for rspec-rails / Rails I believe, it's basically treated as a string and normally invisible to the end user.

File: spec/integration/output_stream_spec.rb (hunk 10 above)
Comment (JonRowe) [inline spec/integration/output_stream_spec.rb]: Spec below was just restored / moved back, not added, but sure.

File: lib/rspec/core/output_wrapper.rb (hunk 7 above)
Comment (myronmarston) [inline lib/rspec/core/output_wrapper.rb]: `puts` is an `IO` instance method, so it gets defined below, so this class will respond to it w/o overriding `respond_to?`.  Unless I'm missing something, the only behavioral change here is that it makes this class claim to respond to methods that it doesn't.  Methods in `IO.instance_methods(false)` are the only ones we proxy to `@output`, and the class already claims to respond to those because they are defined with `define_method` instead of method missing.  OTOH, extra methods that `@output` responds to (such as `StringIO#string`), which are not part of `IO.instance_methods(false)` are not methods that `OutputWrapper` responds to (since we are not using `method_missing`) but that this `respond_to?` implementation will falsely claim that we respond to.

Does that make sense?

File: lib/rspec/core/configuration.rb (hunk 2 above)
Comment (myronmarston) [inline lib/rspec/core/configuration.rb]: sounds good.

File: lib/rspec/core/output_wrapper.rb [hunk 11]
```diff
@@ -3,17 +3,25 @@ module Core
...
         @output = output
       end
 
+      def respond_to?(name, priv=false)
+        output.respond_to?(name, priv)
+      end
+
+      def method_missing(name, *args, &block)
+        output.method_missing(name, *args, &block)
+      end
```
Comment (myronmarston) [inline lib/rspec/core/output_wrapper.rb]: Ah, I thought we didn't define `method_missing` here.  What's the reason to do that?  Do we need to proxy any methods that are not part of the `IO.instance_methods` interface?  (I thought we didn't).

File: lib/rspec/core/output_wrapper.rb (hunk 7 above)
Comment (JonRowe) [inline lib/rspec/core/output_wrapper.rb]: The problem is the inverse case, because this may be a string it will say it responds to puts when it doesn't.

File: lib/rspec/core/output_wrapper.rb (hunk 11 above)
Comment (JonRowe) [inline lib/rspec/core/output_wrapper.rb]: I added it to fix the edge-case you highlighted,

File: lib/rspec/core/output_wrapper.rb (hunk 7 above)
Comment (myronmarston) [inline lib/rspec/core/output_wrapper.rb]: How can `@output` be a string?  I thought we handled strings before wrapping it in this by opening the file and passing the file to this.

This class is specifically designed to wrap objects that implement Ruby's `IO` interface.  IMO, it's a bad idea to let it wrap any other kind of object (like a string).

Comment (JonRowe) [inline lib/rspec/core/output_wrapper.rb]: I moved the strings back to the formatter loader because it's simpler and handles all the existing behaviour, this object is just meant to be a persistent handler to the configured output, I'm happy to drop the IO methods and just use plain old method_missing, it's just that defined methods are quicker

File: lib/rspec/core/output_wrapper.rb (hunk 11 above)
Comment (myronmarston) [inline lib/rspec/core/output_wrapper.rb]: I understand why this is needed now.  But I think it's undesirable to have an `OutputWrapper` class that we specifically have designed to wrap non-output objects.  It's an _output_ wrapper, so it's far less confusing to limit it to only wrapping output objects, and then we don't have to mess with this `respond_to?` and `method_missing` weirdness.

Here's one way to do it:

046179ca43bf96964a51ff645a5007d650e6c742

That passes locally for me.

Comment (JonRowe) [inline lib/rspec/core/output_wrapper.rb]: Breaks the deprecation stream support for string filenames (which was publicly documented beforehand), and the support for adding formatters with string filenames which was implicitly allowed beforehand (some even had the option of multiple string to file names, which is what the map is for in formatter loadeD)

Comment (JonRowe) [inline lib/rspec/core/output_wrapper.rb]: An issue here is we have behaviour that is tested with a minimal cross section, so it looks like it's working but we don't really have enough test cases for all of our supported behaviour.

Comment (JonRowe) [inline lib/rspec/core/output_wrapper.rb]: I initially did exactly what you propose, but because of the wackyness with `deprecation_stream` not liking being an output wrapper abandoned it in favour of restoring the old behaviour.

Comment (myronmarston) [inline lib/rspec/core/output_wrapper.rb]: OK, I can see there's more complexity here I haven't been aware of.

Let's go with what you have (once you get it green), but can you leave a `TODO` command to revisit?  I think it's quite confusing to use both `define_method` for all intended proxy methods and also use `method_missing`.

**Comment by myronmarston** [comment]
LGTM! Merged.
